Split logic to undo txin's off DisconnectBlock.

domob1812 opened the pull request:
Instead, create a separate function that applies the undo operation of a CTxInUndo object onto a CCoinsViewCache.  This method is used from DisconnectBlock.

This is a rebased version of https://github.com/bitcoin/bitcoin/pull/5192.  I had to re-submit it, since the old pull request does no longer pick up changes.  The reason for that is that namecoin/namecoin was changed to be no longer a fork of bitcoin/bitcoin, and the old pull request was done via domob1812/namecoin (forked from namecoin/namecoin).

fanquake commented:
@sipa Care to re-review this?

domob1812 commented:
Rebased.  Can this be merged, or is there anything I still have to do?

TheBlueMatt commented:
utACK

sipa commented on src/main.cpp:
```diff
@@ -1588,6 +1588,39 @@ bool UndoReadFromDisk(CBlockUndo& blockundo, const CDiskBlockPos& pos, const uin
+    if (undo.nHeight != 0) {
+        // undo data contains height: this is the last output of the prevout tx being spent
+        if (!coins->IsPruned())
+            fClean = fClean && error("DisconnectBlock(): undo data overwriting existing transaction");
```
Change the DisconnectBlock() to ApplyTxInUndo() (or even better, replace it with %s: with __FILE__).

sipa commented:
utACK apart from the above nit

domob1812 committed:
Split logic to undo txin's off DisconnectBlock.

Instead, create a separate function that applies the undo operation of a
CTxInUndo object onto a CCoinsViewCache.  This method is used from
DisconnectBlock.

domob1812 commented:
Fixed the nit.

laanwj commented:
ACK, verfied code moves only apart from some error() message changes.
